aptos-move/mvhashmap/src/versioned_aggregators.rs

```diff
@@ -0,0 +1,708 @@
+        use AggregatorEntry::*;
+
+        self.versioned_map
+            .range(0..next_idx_to_commit)
+            .next_back()
+            .map_or(
+                self.base_value.ok_or(MVAggregatorsError::NotFound),
+                |(_, entry)| match &**entry {
+                    Value(v, _) => Ok(*v),
+                    Snapshot(_, _) | Delta(_) => unreachable!(
+                        "Snapshot or Delta entries may not exist for committed txn indices"
+                    ),
+                    Estimate(_) => unreachable!("Committed entry may not be an Estimate"),
+                },
+            )
+    }
+
+    // Traverse down from txn_idx and accumulate encountered deltas until resolving it to
+    // a value or return an error. Errors of not finding a value to resolve to take
+    // precedence over a DeltaApplicationError.
+    fn apply_delta_suffix(
+        &self,
+        iter: &mut dyn DoubleEndedIterator<Item = (&TxnIndex, &CachePadded<AggregatorEntry>)>,
+        delta: DeltaOp,
+    ) -> Result<u128, MVAggregatorsError> {
+        use AggregatorEntry::*;
+
+        let mut accumulator: Result<DeltaOp, ()> = Ok(delta);
+        while let Some((idx, entry)) = iter.next_back() {
+            let delta = match &**entry {
+                Value(v, _) => {
+                    return accumulator
+                        .map_err(|_| MVAggregatorsError::DeltaApplicationFailure)
+                        .and_then(|a| {
+                            // Apply accumulated delta to resolve the aggregator value.
+                            a.apply_to(*v)
+                                .map_err(|_| MVAggregatorsError::DeltaApplicationFailure)
+                        });
+                },
+                Delta(delta) => *delta,
+                Snapshot(_, _) => {
+                    unreachable!("Snapshots and Deltas may not exist for the same Aggregator ID")
+                },
+                Estimate(bypass) => {
+                    if !self.read_estimate_deltas || bypass.is_none() {
+                        // We must wait on Estimates, or a bypass isn't available.
+                        return Err(MVAggregatorsError::Dependency(*idx));
+                    }
+                    let (bypass_delta, maybe_bypass_id) = bypass.expect("None already handled");
+                    assert_none!(
+                        maybe_bypass_id,
+                        "Bypass previously stored for Aggregator ID that contains deltas"
+                    );
+                    bypass_delta
+                },
+            };
+
+            accumulator = accumulator.and_then(|mut a| {
+                // Read hit a delta during traversing the block and aggregating
+                // other deltas. Merge two deltas together. If Delta application
+                // fails, we record an error, but continue processing (to e.g.
+                // account for the case when the aggregator was deleted).
```

Comment:
can aggregator be deleted any more?

Reply:
uh, great point. They can be "deleted" in a way I suppose (say swap & drop), but it shouldn't be a concern to us here, we track by ID and we don't have an explicit deletion flag - and there is nothing to delete from storage at an aggregator granularity.
Simplified the code & updated comments.